### PR TITLE
[master] make odoo.addons a namespace package

### PR DESCRIPTION
**Pull request by sbidoul**
Description of the issue/feature this PR addresses:

make `odoo.addons` behave as a python namespace package, as described in the setuptools documentation http://setuptools.readthedocs.io/en/latest/setuptools.html#namespace-packages

Current behavior before PR:

odoo.addons and openerp.addons are not real python namespace packages and importing from them is made possible by python import hooks which look at an explicit path provided in odoo configuration options.

This has two major drawbacks:
- standard python tools (static code analysis tools, ide code completion) do not work correctly
- it is difficult to package and distribute and install odoo addons

Desired behavior after PR is merged:

When odoo.addons is a namespace package:
- there is no need for import hooks
- packaging and distribution work with standard tools
- static code analysis tools and ide's code completion just works 
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by rco-odoo** [comment]
@sbidoul this looks like a nice trick. I didn't know namespace packages.

However, I am afraid it is not possible for Odoo 10. I recently renamed the directory `openerp` into `odoo` (79331ec09bfff3792040ad307a8d0631236ff5e3), such that `import odoo` no longer needs an import hook. The consequence is that the directory `odoo/addons` is now real, and contains the module `base` and a few others. As such, we cannot convert it to a namespace package.

We are open to suggestions for making `odoo.addons` a namespace package. Maybe moving all the code currently inside `odoo/addons` somewhere else? Right now I don't know what would work best.

**Comment by sbidoul** [comment]
@rco-odoo hey, the renaming is excellent news! It should now be very very easy to make odoo.addons a namespace package. No need at all to move the content of odoo/addons elsewhere.

Adding
```python
import pkg_resources
pkg_resources.declare_namespace(__name__)
```
in `odoo/addons/__init__.py` should do the trick.

If that works, do you agree in principle with the automatic extension of the addons-path like in https://github.com/odoo/odoo/pull/13237/files#diff-a574457167ec463fe510488c27d95214R146 ?

I'll update my PR and do some testing.

**Comment by sbidoul** [comment]
@rco-odoo BTW, if you ask me, I'd bet the community would not mind if old idioms stop working in 10. So you could vastly simplify the code and remove more hooks. Perhaps you should poll the community if all this can be dropped:

```python
import openerp 
import openerp.addons
import crm  # without odoo.addons in front
```

**Comment by sbidoul** [comment]
@rco-odoo here you go, easy, nice and clean.

The first two commits are from #13413 (setup.py cleanup and move odoo.py out of the way to avoid conflict with the odoo package)

File: odoo/tools/config.py [hunk 1]
```diff
@@ -416,7 +416,7 @@ def die(cond, msg):
             elif isinstance(self.options[arg], basestring) and self.casts[arg].type in optparse.Option.TYPE_CHECKER:
                 self.options[arg] = optparse.Option.TYPE_CHECKER[self.casts[arg].type](self.casts[arg], arg, self.options[arg])
 
-        self.options['root_path'] = os.path.abspath(os.path.expanduser(os.path.expandvars(os.path.dirname(odoo.__file__))))
+        self.options['root_path'] = os.path.abspath(os.path.expanduser(os.path.expandvars(os.path.join(os.path.dirname(__file__), '..'))))
```
Comment (rco-odoo) [inline odoo/tools/config.py]: I don't understand why this would be more reliable, provided we move/rename `odoo.py`.

**Comment by rco-odoo** [comment]
@sbidoul having `setup.py` create a command `odoo` is really nice.
But we don't like the idea of moving `odoo.py` in a somewhat hidden place.
We would like to have the possibility to launch Odoo with a script without the need of `setup.py`.
So we propose to rename `odoo.py` to `odoo-bin`, and keep it next to the package `odoo`.

What do you think?

**Comment by sbidoul** [comment]
@rco-odoo it's best practice to use the `console_script` entry point for installing scripts. It works multi platform (creates a .bat on windows for instance), and `pip uninstall` removes them. In my understanding `script` is somewhat deprecated as a setup keyword.

BTW, openerp-server and openerp-server-gevent should be installed the same way.

I fully understand the need to have the odoo script at the source root. 

I suggest to do both
- move the meat of the script to odoo/cli
- install it with a console_script entry point
- have odoo-bin as a very shallow script at the source root, which calls the same meat in odoo/cli

File: odoo/tools/config.py (hunk 1 above)
Comment (sbidoul) [inline odoo/tools/config.py]: With namespace package, the same `odoo` package can have source in different places on the file system. For instance it can be the main odoo package but also the root of some random odoo/addon/someaddon residing elsewhere on disk. 

I think @lmignon has met that issue when doing preliminary testing last year.

`dirname(__file__)/..` OTOH is sure to point to the right directory in all cases. So it's safer.

Comment (lmignon) [inline odoo/tools/config.py]: @sbidoul you're right!

**Comment by antonylesuisse** [comment]
openerp-server and openerp-server-gevent  should be removed @sle-odoo i think you have a pr for that.

**Comment by antonylesuisse** [comment]
Ok it's done in 76c7041c5bb309844008c188683b7a55f186fb7b

**Comment by rco-odoo** [comment]
@sbidoul sorry for my unclear message. I actually meant to have both mechanisms: the console script and the python script in the root directory. Could you adapt your PR as you suggested?

File: odoo/tools/config.py (hunk 1 above)
Comment (rco-odoo) [inline odoo/tools/config.py]: Thanks for the explanation!

**Comment by rco-odoo** [comment]
@sbidoul by the way, can you retarget the pull request to `saas-13` instead of `master`? It will be forward-ported to `master` anyway. Thanks!

**Comment by sbidoul** [comment]
@rco-odoo done.

Note there are still hardwired references to `odoo.py` all over the place. I'd be happy to change them, but I prefer to let you guys go through them, because I'm not familiar with all of these use cases (especially those related to your own dev workflow).

The part related to the odoo.py removal is also on #13413 if you prefer to work on it independently of the ns package thing.

I gave you access to both PR with the new github feature.

File: setup.py [hunk 2]
```diff
@@ -129,9 +132,14 @@ def py2exe_options():
     author_email=author_email,
     classifiers=filter(None, classifiers.split('\n')),
     license=license,
-    scripts=['openerp-server', 'odoo.py'],
+    scripts=['openerp-server'],
+    entry_points={
+        'console_scripts': [
+            'odoo-bin=odoo.cli.odoobin:main',
```
Comment (odony) [inline setup.py]: @sbidoul @rco-odoo Shouldn't this be `odoo=odoo.cli.odoobin:main`? Our idea would be that when you do a local installation of Odoo you get to use the simplest `odoo` command everywhere, à la `rails`. This way you can conveniently run `odoo shell -d test` from anywhere.
On the other hand if you only do a `git clone` you'll still be able to directly run odoo with the same parameters/API, using `./odoo-bin` instead of `odoo` (that's the closest we can do, I guess). E.g `./odoo-bin shell -d test`

Comment (sbidoul) [inline setup.py]: @odony any option is good for me. 

I actually ended up doing a few commits on this topic because I needed to update setup.py and get odoo.py out of the way because it conflicts with the odoo package.

Actually this is somewhat unrelated with the title of this PR and is best discussed on #13413

Comment (sbidoul) [inline setup.py]: I added some commits in #13413 to explore a bit further.

**Comment by sbidoul** [comment]
@rco-odoo @odony I think tried to go to far with this `odoo.py` removal. I now rebased on the simplest possible way to do it (in #13413).

**Comment by sbidoul** [comment]
@rco-odoo @odony @antonylesuisse what do you think of this version?

**Comment by rco-odoo** [comment]
@sle-odoo will take this over.

**Comment by sbidoul** [comment]
@sle-odoo let me know if you want to discuss it. I'm happy to help and I'm keen to having this working before Odoo Experience so I can showcase it in my talk.

**Comment by sbidoul** [comment]
merged at https://github.com/odoo/odoo/commit/389c2ba97ba83dea01949d2a3cad295bd710d9ef
